### PR TITLE
 Hotfix/increase genesis balances

### PR DESCRIPTION
**Pull request by kstdl**
Related PR that should be merged first https://github.com/Taraxa-project/taraxa-evm/pull/33

### CODE REVIEW

File: libraries/cli/src/config.cpp [hunk 1]
```diff
@@ -160,6 +160,22 @@ Config::Config(int argc, const char* argv[]) {
     Json::Value config_json = Tools::readJsonFromFile(config);
     Json::Value wallet_json = Tools::readJsonFromFile(wallet);
 
+    auto override = [&]() {
```
Comment (JakubFornadel) [inline libraries/cli/src/config.cpp:163]: override is C++ keyword(identifier with special meaning), add some prefix or adjust somehow the name for this function

File: libraries/cli/src/config.cpp [hunk 2]
```diff
@@ -215,6 +230,22 @@ bool Config::nodeConfigured() { return node_configured_; }
 
 FullNodeConfig Config::getNodeConfiguration() { return node_config_; }
 
+void Config::addNewHardforks(Json::Value& config, const Json::Value& default_config) {
+  // network_id is exactly the same thing as chain_id. So get it from config
+  auto& new_hardforks_json = default_config["chain_config"]["final_chain"]["state"]["hardforks"];
+  auto& local_hardforks_json = config["chain_config"]["final_chain"]["state"]["hardforks"];
+
+  if (local_hardforks_json.isNull()) {
+    local_hardforks_json = new_hardforks_json;
```
Comment (JakubFornadel) [inline libraries/cli/src/config.cpp:239]: missing return here ?

File: libraries/cli/include/cli/testnet_config.hpp [hunk 3]
```diff
@@ -102,6 +102,7 @@ const char *testnet_json = R"foo({
           "deposit_delay": "0x5",
           "withdrawal_delay": "0x5",
           "eligibility_balance_threshold": "0xf4240",
+          "coins_per_vote": "0xf4240",
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/testnet_config.hpp:105]: why is this called coins_per_vote and not eligibility_balance_step ? Would that be better ?

File: libraries/config/src/chain_config.cpp [hunk 4]
```diff
@@ -52,6 +52,7 @@ decltype(ChainConfig::predefined_) const ChainConfig::predefined_([] {
     cfg.final_chain.state.genesis_balances[root_node_addr] = 9007199254740991;
     auto& dpos = cfg.final_chain.state.dpos.emplace();
     dpos.eligibility_balance_threshold = 1000000000;
+    dpos.coins_per_vote = 1000000000;
```
Comment (JakubFornadel) [inline libraries/config/src/chain_config.cpp:55]: isnt this supposed to be 100k ?

File: libraries/cli/include/cli/default_config.hpp [hunk 5]
```diff
@@ -202,8 +203,11 @@ const char *default_json = R"foo({
           "disable_nonce_check": true
         },
         "genesis_balances": {
-          "0274cfffea9fa850e54c93a23042f12a87358a82": "0x141e8d17",
-          "111f91441efc8c6c0edf6534970cc887e2fabaa8": "0x24048ce3d"
+          "0274cfffea9fa850e54c93a23042f12a87358a82": "0x141e8d17000000000000000000",
+          "111f91441efc8c6c0edf6534970cc887e2fabaa8": "0x24048ce3d000000000000000000"
+        },
+        "hardforks": {
+          "fix_genesis_hardfork_block_num": "0x12c"
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/default_config.hpp:210]: so dont we want to start using tip_<NUM> as `Taraxa Improvement Proposal` naming for hardforks ?

File: libraries/core_libs/consensus/include/final_chain/final_chain.hpp [hunk 6]
```diff
@@ -18,9 +18,11 @@ class FinalChain {
 
  protected:
   util::EventEmitter<std::shared_ptr<FinalizationResult>> const block_finalized_emitter_{};
+  util::EventEmitter<uint64_t> const block_applying_emitter_{};
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/include/final_chain/final_chain.hpp:21]: who is using this emitter ? I know what event it emits, but who and when needs it ?

File: libraries/core_libs/consensus/src/final_chain/state_api.cpp [hunk 7]
```diff
@@ -225,6 +234,10 @@ bool StateAPI::dpos_is_eligible(EthBlockNumber blk_num, addr_t const& addr) cons
   return ret;
 }
 
+u256 StateAPI::get_staking_balance(EthBlockNumber blk_num, addr_t const& addr) const {
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/final_chain/state_api.cpp:237]: const addr_t&

File: libraries/core_libs/network/rpc/Test.cpp [hunk 8]
```diff
@@ -281,7 +281,7 @@ Json::Value Test::get_all_nodes() {
   return res;
 }
 
-Json::Value Test::get_vote_weight(const Json::Value &param1) {
+Json::Value Test::get_coins_per_vote(const Json::Value &param1) {
```
Comment (JakubFornadel) [inline libraries/core_libs/network/rpc/Test.cpp:284]: I am totally lost why should be this method renamed to get_coins_per_vote ? It still returns vote_weight doesnt it ?

File: libraries/cli/src/config.cpp [hunk 9]
```diff
@@ -215,6 +230,22 @@ bool Config::nodeConfigured() { return node_configured_; }
 
 FullNodeConfig Config::getNodeConfiguration() { return node_config_; }
 
+void Config::addNewHardforks(Json::Value& config, const Json::Value& default_config) {
+  // network_id is exactly the same thing as chain_id. So get it from config
```
Comment (MatusKysel) [inline libraries/cli/src/config.cpp:234]: is this comment still relevant? because I do not understand it

File: libraries/core_libs/network/rpc/Test.cpp (hunk 8 above)
Comment (kstdl) [reply libraries/core_libs/network/rpc/Test.cpp:284]: It is just rename issue. I accidentally changed it when I was renaming that new config field

File: libraries/cli/include/cli/testnet_config.hpp (hunk 3 above)
Comment (kstdl) [reply libraries/cli/include/cli/testnet_config.hpp:105]: Maybe `vote_eligibility_balance_step` is better?

File: libraries/core_libs/consensus/src/final_chain/final_chain.cpp [hunk 10]
```diff
@@ -160,14 +161,14 @@ class FinalChainImpl final : public FinalChain {
...
     db_->commitWriteBatch(batch, db_opts_w_);
+    block_finalized_emitter_.emit(result);
     state_api_.transition_state_commit();
     {
       unique_lock l(last_block_mu_);
       last_block_ = blk_header;
     }
     num_executed_dag_blk_ = num_executed_dag_blk;
     num_executed_trx_ = num_executed_trx;
-    block_finalized_emitter_.emit(result);
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/final_chain/final_chain.cpp:170]: you sure about this change? shouldn't we wait for `transition_state_commit`

File: libraries/core_libs/network/rpc/Test.cpp (hunk 8 above)
Comment (MatusKysel) [inline libraries/core_libs/network/rpc/Test.cpp:284]: This should no be changed, it returns binomial distribution of weigh of vote per specific period and step

File: libraries/core_libs/node/include/node/node.hpp [hunk 11]
```diff
@@ -118,4 +118,14 @@ class FullNode : public std::enable_shared_from_this<FullNode> {
   void rebuildDb();
 };
 
+class HardforkExecutors {
```
Comment (MatusKysel) [inline libraries/core_libs/node/include/node/node.hpp:121]: when this is used?

File: libraries/cli/include/cli/default_config.hpp (hunk 5 above)
Comment (kstdl) [reply libraries/cli/include/cli/default_config.hpp:210]: Did we agreed on how we do it? We could crate TIP describing this hardfork and name it `tip1_block_num`. But we should have that TIP first

File: libraries/core_libs/node/src/node.cpp [hunk 12]
```diff
@@ -185,6 +185,30 @@ void FullNode::start() {
           }
         },
         *rpc_thread_pool_);
+
+    // Subscription to process hardforks
+    final_chain_->block_applying.subscribe([&](uint64_t block_num) {
```
Comment (JakubFornadel) [inline libraries/core_libs/node/src/node.cpp:190]: What is the advantage to use emitter and trigger the hf function here vs not using any emitter and trigger hf function in `finalize` function ?

File: libraries/core_libs/network/rpc/Test.cpp (hunk 8 above)
Comment (kstdl) [reply libraries/core_libs/network/rpc/Test.cpp:284]: It is just rename issue :)
https://github.com/Taraxa-project/taraxa-node/pull/1407#discussion_r787529065

File: libraries/cli/src/config.cpp (hunk 9 above)
Comment (kstdl) [reply libraries/cli/src/config.cpp:234]: Yeah, it should not be there

File: libraries/config/src/chain_config.cpp (hunk 4 above)
Comment (kstdl) [reply libraries/config/src/chain_config.cpp:55]: It is used for tests. So to not mess values and tests it should be the same as `eligibility_balance_threshold`

File: libraries/core_libs/node/include/node/node.hpp (hunk 11 above)
Comment (kstdl) [reply libraries/core_libs/node/include/node/node.hpp:121]: It is not used right now. I will remove it

File: libraries/core_libs/consensus/src/final_chain/final_chain.cpp (hunk 10 above)
Comment (kstdl) [reply libraries/core_libs/consensus/src/final_chain/final_chain.cpp:170]: It works fine, but I just forgot to move it back after tries to apply hardfork also on that event(to not create new one). Will move it back

File: libraries/cli/include/cli/default_config.hpp (hunk 5 above)
Comment (MatusKysel) [reply libraries/cli/include/cli/default_config.hpp:210]: maybe not even _num this is how ETH is doing it 
```    "homesteadBlock": 0,
    "eip150Block": 0,
    "eip155Block": 0,
    "eip158Block": 0,
    "byzantiumBlock": 0,
    "constantinopleBlock": 0,
    "petersburgBlock": 0,
```

File: libraries/cli/include/cli/testnet_config.hpp (hunk 3 above)
Comment (JakubFornadel) [reply libraries/cli/include/cli/testnet_config.hpp:105]: yeah even better, I skipped "vote_" so it has the smae form as eligibility_balance_threshold, which is missing vote_ prefix too

File: libraries/cli/include/cli/default_config.hpp (hunk 5 above)
Comment (JakubFornadel) [reply libraries/cli/include/cli/default_config.hpp:210]: I remeber @justinsnapp agreed on calling it tip, not sure about the others though ?

File: libraries/core_libs/node/src/node.cpp (hunk 12 above)
Comment (kstdl) [reply libraries/core_libs/node/src/node.cpp:190]: We need to apply that before block is finalized and state commited, so I created one another emitter. 
Also it is in the node to be more generic thing. Because we can not change a lot of things from final_chain
Also I just realised that I forgot about config updation. Config file should be updated from that part of code also, not just inmemory config. So it is another one argument for it to be here I guess

File: tests/full_node_test.cpp [hunk 13]
```diff
@@ -1525,6 +1531,96 @@ TEST_F(FullNodeTest, chain_config_json) {
...
+  cli::Config::addNewHardforks(config, default_json);
+  EXPECT_EQ(state_cfg["hardforks"], default_hardforks);
+}
+
+TEST_F(FullNodeTest, fix_genesis_fork_block_is_zero) {
+  auto node_cfgs = make_node_cfgs(1);
+  auto &cfg = node_cfgs.front().chain.final_chain;
+  cfg.state.hardforks.fix_genesis_fork_block = 0;
+  auto node = launch_nodes(node_cfgs).front();
+}
```
Comment (gaoqi8649) [inline tests/full_node_test.cpp:1557]: There is no any checking in the unit test

File: for_devs/local-net [hunk 14]
```diff
@@ -158,11 +158,13 @@ def config(binary, boot_nodes, rpc_nodes):
                 del genesis_balances[genesis_balances_faucet_address]
                 genesis_balances[faucet_public_address] = genesis_balances_faucet_value
 
+                genesis_state = genesis_state[list(genesis_state.keys())[0]]
+                dpos_config_balance = genesis_state[list(genesis_state.keys())[0]]
```
Comment (MatusKysel) [inline for_devs/local-net:162]: what is the difference between `genesis_state`  and `dpos_config_balance` ?

File: libraries/cli/include/cli/default_config.hpp [hunk 15]
```diff
@@ -204,6 +205,9 @@ const char *default_json = R"foo({
         "genesis_balances": {
           "0274cfffea9fa850e54c93a23042f12a87358a82": "0x141e8d17",
           "111f91441efc8c6c0edf6534970cc887e2fabaa8": "0x24048ce3d"
+        },
+        "hardforks": {
+          "fix_genesis_fork_block": "0x64"
```
Comment (MatusKysel) [inline libraries/cli/include/cli/default_config.hpp:210]: why can we use normal integers instead of hex? `"fix_genesis_fork_block": 100`

File: tests/state_api_test.cpp [hunk 16]
```diff
@@ -178,7 +180,8 @@ TEST_F(StateAPITest, dpos_integration) {
   EXEC_AND_CHECK({});
 }
 
-TEST_F(StateAPITest, eth_mainnet_smoke) {
+// disabled because of changing balances hardfork
+TEST_F(StateAPITest, DISABLED_eth_mainnet_smoke) {
```
Comment (MatusKysel) [inline tests/state_api_test.cpp:184]: can you enable it by disabling you HF?

File: for_devs/local-net (hunk 14 above)
Comment (kstdl) [reply for_devs/local-net:162]: If you look closely in the first line we are overwriting `genesis_state`, so the values are different. 
First one is mapping address -> balance, but second one is just balance to save value from config. 
That done to save value from the config, but change addresses. Before it was just setting to eligibility_balance_threshold

File: libraries/cli/include/cli/default_config.hpp (hunk 15 above)
Comment (kstdl) [reply libraries/cli/include/cli/default_config.hpp:210]: Currently we couldn't use decimal ints, because we expecting hex value there. But it is not hard to implement feature as I know

File: tests/state_api_test.cpp (hunk 16 above)
Comment (kstdl) [reply tests/state_api_test.cpp:184]: Yeah, it should work. I will try

File: libraries/config/include/config/hardfork.hpp [hunk 17]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include <libdevcore/CommonJS.h>
+
+#include "common/encoding_rlp.hpp"
+
+struct Hardforks {
+  uint64_t fix_genesis_fork_block = 0;
+  void processFixGenesisHardfork() const;
```
[the remark this answers is not in the document]
Comment (kstdl) [reply libraries/config/include/config/hardfork.hpp:9]: removed

File: libraries/cli/include/cli/default_config.hpp [hunk 18]
```diff
@@ -134,6 +134,10 @@ const char *default_json = R"foo({
           {
             "name": "SUMMARY",
             "verbosity": "INFO"
+          },
+          {
+            "name": "SORT_MGR",
+            "verbosity": "DEBUG"
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/default_config.hpp:140]: this change does not seem to be related to genesis balances hardfork ?

File: libraries/cli/include/cli/default_config.hpp (hunk 15 above)
Comment (MatusKysel) [reply libraries/cli/include/cli/default_config.hpp:214]: but I saw this `json["homestead_block"] = dev::toJS(obj.homestead_block);` so we can parse it same way as eth does. Or I think you can leave it for now, I have created task, to switch everything to readable numbers

**Review by MatusKysel — APPROVED** [review]

**Review by JakubFornadel — APPROVED** [review]

**Review by mfrankovi — APPROVED** [review]

File: libraries/cli/include/cli/default_config.hpp (hunk 18 above)
Comment (kstdl) [reply libraries/cli/include/cli/default_config.hpp:140]: I added it to get some logs during PRnet testing, so I will just remove it after

File: libraries/cli/include/cli/default_config.hpp (hunk 5 above)
Comment (kstdl) [reply libraries/cli/include/cli/default_config.hpp]: Agreed there that to make TIP name we should have some TIP description in a proper place for it. As we didn't have one it makes sense to call it with a "speaking name"
